File: FirebaseAuth/Tests/Sample/Sample/SettingsViewController.m
/*
 * Copyright 2019 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "SettingsViewController.h"

#import <objc/runtime.h>

#import "AppManager.h"
#import "FirebaseAuth/Sources/Auth/FIRAuth_Internal.h"
#import "FirebaseAuth/Sources/SystemService/FIRAuthAPNSToken.h"
#import "FirebaseAuth/Sources/SystemService/FIRAuthAPNSTokenManager.h"
#import "FirebaseAuth/Sources/SystemService/FIRAuthAppCredential.h"
#import "FirebaseAuth/Sources/SystemService/FIRAuthAppCredentialManager.h"
#import <FirebaseCore/FIRApp.h>
#import <FirebaseCore/FIROptions.h>
#import <FirebaseCore/FIRVersion.h>
#import <FirebaseAuth/FirebaseAuth.h>
#import "StaticContentTableViewManager.h"
#import "UIViewController+Alerts.h"

/** @var kIdentityToolkitRequestClassName
    @brief The class name of Identity Toolkit requests.
 */
static NSString *const kIdentityToolkitRequestClassName = @"FIRIdentityToolkitRequest";

/** @var kSecureTokenRequestClassName
    @brief The class name of Secure Token Service requests.
 */
static NSString *const kSecureTokenRequestClassName = @"FIRSecureTokenRequest";

/** @var kIdentityToolkitSandboxHost
    @brief The host of Identity Toolkit sandbox server.
 */
static NSString *const kIdentityToolkitSandboxHost = @"staging-www.sandbox.googleapis.com";

/** @var kSecureTokenSandboxHost
    @brief The host of Secure Token Service sandbox server.
 */
static NSString *const kSecureTokenSandboxHost = @"staging-securetoken.sandbox.googleapis.com";

/** @var kGoogleServiceInfoPlists
    @brief a C-array of plist file base names of Google service info to initialize FIRApp.
 */
static NSString *const kGoogleServiceInfoPlists[] = {
  @"GoogleService-Info",
  @"GoogleService-Info_multi"
};

/** @var kSharedKeychainAccessGroup
    @brief The shared keychain access group for testing.
 */
static NSString *const kSharedKeychainAccessGroup = @"com.google.firebase.auth.keychainGroup1";

/** @var gAPIEndpoints
    @brief List of API Hosts by request class name.
 */
static NSDictionary<NSString *, NSArray<NSString *> *> *gAPIHosts;

/** @var gFirebaseAppOptions
    @brief List of FIROptions.
 */
static NSArray<FIROptions *> *gFirebaseAppOptions;

/** @protocol RequestClass
    @brief A de-facto protocol followed by request class objects to access its API host.
 */
@protocol RequestClass <NSObject>
- (NSString *)host;
- (void)setHost:(NSString *)host;
@end

/** @fn requestHost
    @brief Retrieves the API host for the request class.
    @param requestClassName The name of the request class.
 */
static NSString *APIHost(NSString *requestClassName) {
  return [(id<RequestClass>)NSClassFromString(requestClassName) host];
}

/** @fn truncatedString
    @brief Truncates a string under a maximum length.
    @param string The original string to be truncated.
    @param length The maximum length of the truncated string.
    @return The truncated string, which is not longer than @c length.
 */
static NSString *truncatedString(NSString *string, NSUInteger length) {
  if (string.length <= length) {
    return string;
  }
  NSUInteger half = (length - 3) / 2;
  return [NSString stringWithFormat:@"%@...%@",
                                    [string substringToIndex:half],
                                    [string substringFromIndex:string.length - half]];
}

@implementation SettingsViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  [self setUpAPIHosts];
  [self setUpFirebaseAppOptions];
  [self loadTableView];
}

- (IBAction)done:(id)sender {
  [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)setUpAPIHosts {
  if (gAPIHosts) {
    return;
  }
  gAPIHosts = @{
    kIdentityToolkitRequestClassName : @[
      APIHost(kIdentityToolkitRequestClassName),
      kIdentityToolkitSandboxHost,
    ],
    kSecureTokenRequestClassName : @[
      APIHost(kSecureTokenRequestClassName),
      kSecureTokenSandboxHost,
    ],
  };
}

- (void)setUpFirebaseAppOptions {
  if (gFirebaseAppOptions) {
    return;
  }
  int numberOfOptions = sizeof(kGoogleServiceInfoPlists) / sizeof(*kGoogleServiceInfoPlists);
  NSMutableArray *appOptions = [[NSMutableArray alloc] initWithCapacity:numberOfOptions];
  for (int i = 0; i < numberOfOptions; i++) {
    NSString *plistFileName = kGoogleServiceInfoPlists[i];
    NSString *plistFilePath = [[NSBundle mainBundle] pathForResource:plistFileName
                                                              ofType:@"plist"];
    FIROptions *options = [[FIROptions alloc] initWithContentsOfFile:plistFilePath];
    [appOptions addObject:options];
  }
  gFirebaseAppOptions = [appOptions copy];
}

- (void)loadTableView {
  NSString *appIdentifierPrefix = NSBundle.mainBundle.infoDictionary[@"AppIdentifierPrefix"];
  NSString *fullKeychainAccessGroup = [appIdentifierPrefix stringByAppendingString:kSharedKeychainAccessGroup];

  __weak typeof(self) weakSelf = self;
  _tableViewManager.contents = [StaticContentTableViewContent contentWithSections:@[
    [StaticContentTableViewSection sectionWithTitle:@"Versions" cells:@[
      [StaticContentTableViewCell cellWithTitle:@"FirebaseAuth"
                                          value:FIRFirebaseVersion()],
    ]],
    [StaticContentTableViewSection sectionWithTitle:@"API Hosts" cells:@[
      [StaticContentTableViewCell cellWithTitle:@"Identity Toolkit"
                                          value:APIHost(kIdentityToolkitRequestClassName)
                                         action:^{
        [weakSelf toggleAPIHostWithRequestClassName:kIdentityToolkitRequestClassName];
      }],
      [StaticContentTableViewCell cellWithTitle:@"Secure Token"
                                          value:APIHost(kSecureTokenRequestClassName)
                                         action:^{
        [weakSelf toggleAPIHostWithRequestClassName:kSecureTokenRequestClassName];
      }],
    ]],
    [StaticContentTableViewSection sectionWithTitle:@"Firebase Apps" cells:@[
      [StaticContentTableViewCell cellWithTitle:@"Active App"
                                          value:[self activeAppDescription]
                                         action:^{
        [weakSelf toggleActiveApp];
      }],
      [StaticContentTableViewCell cellWithTitle:@"Default App"
                                          value:[self projectIDForAppAtIndex:0]
                                         action:^{
        [weakSelf toggleProjectForAppAtIndex:0];
      }],
      [StaticContentTableViewCell cellWithTitle:@"Other App"
                                          value:[self projectIDForAppAtIndex:1]
                                         action:^{
        [weakSelf toggleProjectForAppAtIndex:1];
      }],
    ]],
    [StaticContentTableViewSection sectionWithTitle:@"Keychain Access Groups" cells:@[
      [StaticContentTableViewCell cellWithTitle:@"Current Access Group"
                                          value:[AppManager auth].userAccessGroup ? [AppManager auth].userAccessGroup : @"[none]"
      ],
      [StaticContentTableViewCell cellWithTitle:@"Default Group"
                                          value:@"[none]"
                                         action:^{
                                           [[AppManager auth] useUserAccessGroup:nil error:nil];
                                           [self loadTableView];
      }],
      [StaticContentTableViewCell cellWithTitle:@"Shared Group"
                                          value:fullKeychainAccessGroup
                                         action:^{
                                           [[AppManager auth] useUserAccessGroup:fullKeychainAccessGroup error:nil];
                                           [self loadTableView];
      }],
    ]],
    [StaticContentTableViewSection sectionWithTitle:@"Phone Auth" cells:@[
      [StaticContentTableViewCell cellWithTitle:@"APNs Token"
                                          value:[self APNSTokenString]
                                         action:^{
        [weakSelf clearAPNSToken];
      }],
      [StaticContentTableViewCell cellWithTitle:@"App Credential"
                                          value:[self appCredentialString]
                                         action:^{
        [weakSelf clearAppCredential];
      }],
    ]],
    [StaticContentTableViewSection sectionWithTitle:@"Language" cells:@[
      [StaticContentTableViewCell cellWithTitle:@"Auth Language"
                                          value:[AppManager auth].languageCode ?: @"[none]"
                                         action:^{
        [weakSelf showLanguageInput];
      }],
      [StaticContentTableViewCell cellWithTitle:@"Use App language" action:^{
        [[AppManager auth] useAppLanguage];
        [weakSelf loadTableView];
      }],
    ]],
    [StaticContentTableViewSection sectionWithTitle:@"Auth Settings" cells:@[
      [StaticContentTableViewCell cellWithTitle:@"Disable App Verification (Phone)"
                                          value:[AppManager auth].settings.
                                              appVerificationDisabledForTesting ? @"Yes" : @"No"
                                         action:^{
        [weakSelf toggleDisableAppVerification];
        [weakSelf loadTableView];
      }],
    ]],
  ]];
}

/** @fn toggleDisableAppVerification
    @brief Toggles the appVerificationDisabledForTesting flag on the current Auth instance.
 */
- (void)toggleDisableAppVerification {
  [AppManager auth].settings.appVerificationDisabledForTesting =
      ![AppManager auth].settings.appVerificationDisabledForTesting;
}

/** @fn toggleAPIHostWithRequestClassName:
    @brief Toggles the host name of the server that handles RPCs.
    @param requestClassName The name of the RPC request class.
 */
- (void)toggleAPIHostWithRequestClassName:(NSString *)requestClassName {
  NSString *currentHost = APIHost(requestClassName);
  NSArray<NSString *> *allHosts = gAPIHosts[requestClassName];
  NSString *newHost = allHosts[([allHosts indexOfObject:currentHost] + 1) % allHosts.count];
  [(id<RequestClass>)NSClassFromString(requestClassName) setHost:newHost];
  [self loadTableView];
}

/** @fn activeAppDescription
    @brief Returns the description for the currently active Firebase app.
 */
- (NSString *)activeAppDescription {
  return [AppManager sharedInstance].active == 0 ? @"[Default]" : @"[Other]";
}

/** @fn toggleActiveApp
    @brief Toggles the active Firebase app for the rest of the application.
 */
- (void)toggleActiveApp {
  AppManager *apps = [AppManager sharedInstance];
  // This changes the FIRAuth instance returned from `[AppManager auth]` to be one that is
  // associated with a different `FIRApp` instance. The sample app uses `[AppManager auth]`
  // instead of `[FIRAuth auth]` almost everywhere. Thus, this statement switches between default
  // and non-default `FIRApp` instances for the sample app to test against.
  apps.active = (apps.active + 1) % apps.count;
  [self loadTableView];
}

/** @fn projectIDForAppAtIndex:
    @brief Returns the Firebase project ID for the Firebase app at the given index.
    @param index The index for the app in the app manager.
    @return The ID of the project.
 */
- (NSString *)projectIDForAppAtIndex:(int)index {
  NSString *APIKey = [[AppManager sharedInstance] appAtIndex:index].options.APIKey;
  for (FIROptions *options in gFirebaseAppOptions) {
    if ([options.APIKey isEqualToString:APIKey]) {
      return options.projectID;
    }
  }
  return @"[none]";
}

/** @fn projectIDForAppAtIndex:
    @brief Returns the Firebase project ID for the Firebase app at the given index.
    @param index The index for the app in the app manager.
    @return The ID of the project.
 */
- (NSString *)keychainAccessGroupAtIndex:(int)index {
  NSArray *array = @[@"123", @"456"];
  return array[index];
}

/** @fn toggleProjectForAppAtIndex:
    @brief Toggles the Firebase project for the Firebase app at the given index by recreating the
        FIRApp instance with different options.
    @param index The index for the app to be recreated in the app manager.
 */
- (void)toggleProjectForAppAtIndex:(int)index {
  NSString *APIKey = [[AppManager sharedInstance] appAtIndex:index].options.APIKey;
  int optionIndex;
  for (optionIndex = 0; optionIndex < gFirebaseAppOptions.count; optionIndex++) {
    FIROptions *options = gFirebaseAppOptions[optionIndex];
    if ([options.APIKey isEqualToString:APIKey]) {
      break;
    }
  }

  FIROptions *options;
  if (index == 0) {
    // For default apps, the next options cannot be `nil`.
    optionIndex = (optionIndex + 1) % gFirebaseAppOptions.count;
    options = gFirebaseAppOptions[optionIndex];
  } else {
    // For non-default apps, `nil` is considered the next options after the last options in the array.
    optionIndex = (optionIndex + 1) % (gFirebaseAppOptions.count + 1);
    if (optionIndex != gFirebaseAppOptions.count) {
      options = gFirebaseAppOptions[optionIndex];
    }
  }
  __weak typeof(self) weakSelf = self;
  [[AppManager sharedInstance] recreateAppAtIndex:index withOptions:options completion:^() {
    dispatch_async(dispatch_get_main_queue(), ^() {
      [weakSelf loadTableView];
    });
  }];
}

/** @fn APNSTokenString
    @brief Returns a string representing APNS token.
 */
- (NSString *)APNSTokenString {
  FIRAuthAPNSToken *token = [AppManager auth].tokenManager.token;
  if (!token) {
    return @"";
  }
  return [NSString stringWithFormat:@"%@(%@)",
                                    truncatedString(token.string, 19),
                                    token.type == FIRAuthAPNSTokenTypeProd ? @"P" : @"S"];
}

/** @fn clearAPNSToken
    @brief Clears the saved app credential.
 */
- (void)clearAPNSToken {
  FIRAuthAPNSToken *token = [AppManager auth].tokenManager.token;
  if (!token) {
    return;
  }
  NSString *tokenType = token.type == FIRAuthAPNSTokenTypeProd ? @"Production" : @"Sandbox";
  NSString *message = [NSString stringWithFormat:@"token: %@\ntype: %@",
                                                 token.string, tokenType];
  [self showMessagePromptWithTitle:@"Clear APNs Token?"
                           message:message
                  showCancelButton:YES
                        completion:^(BOOL userPressedOK, NSString *_Nullable userInput) {
    if (userPressedOK) {
      [AppManager auth].tokenManager.token = nil;
      [self loadTableView];
    }
  }];
}

/** @fn appCredentialString
    @brief Returns a string representing app credential.
 */
- (NSString *)appCredentialString {
  FIRAuthAppCredential *credential = [AppManager auth].appCredentialManager.credential;
  if (!credential) {
    return @"";
  }
  return [NSString stringWithFormat:@"%@/%@",
                                    truncatedString(credential.receipt, 13),
                                    truncatedString(credential.secret, 13)];
}

/** @fn clearAppCredential
    @brief Clears the saved app credential.
 */
- (void)clearAppCredential {
  FIRAuthAppCredential *credential = [AppManager auth].appCredentialManager.credential;
  if (!credential) {
    return;
  }
  NSString *message = [NSString stringWithFormat:@"receipt: %@\nsecret: %@",
                                                 credential.receipt, credential.secret];
  [self showMessagePromptWithTitle:@"Clear App Credential?"
                           message:message
                  showCancelButton:YES
                        completion:^(BOOL userPressedOK, NSString *_Nullable userInput) {
    if (userPressedOK) {
      [[AppManager auth].appCredentialManager clearCredential];
      [self loadTableView];
    }
  }];
}

/** @fn showLanguageInput
    @brief Show language code input field.
 */
- (void)showLanguageInput {
  [self showTextInputPromptWithMessage:@"Enter Language Code For Auth:"
                       completionBlock:^(BOOL userPressedOK, NSString *_Nullable languageCode) {
    if (!userPressedOK) {
      return;
    }
    [AppManager auth].languageCode = languageCode.length ? languageCode : nil;
    [self loadTableView];
  }];
}

@end

File: FirebaseAuth/Tests/SampleSwift/AuthenticationExample/ViewControllers/SettingsViewController.swift
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

@testable import FirebaseAuth
import FirebaseCore
import UIKit

/// Namespace for performable actions on a Auth Settings view
enum SettingsAction: String {
  case toggleIdentityTokenAPI = "Identity Toolkit"
  case toggleSecureTokenAPI = "Secure Token"
  case toggleActiveApp = "Active App"
  case toggleAccessGroup = "Current Access Group"
  case toggleAPNSToken = "APNs Token"
  case toggleAppCredential = "App Credential"
  case setAuthLanguage = "Auth Language"
  case useAppLanguage = "Use App Language"
  case togglePhoneAppVerification = "Disable App Verification (Phone)"
}

class SettingsViewController: UIViewController, DataSourceProviderDelegate {
  var dataSourceProvider: DataSourceProvider<AuthSettings>!

  var tableView: UITableView { view as! UITableView }

  private var _settings: AuthSettings?
  var settings: AuthSettings? {
    get { AppManager.shared.auth().settings }
    set { _settings = newValue }
  }

  // MARK: - UIViewController Life Cycle

  override func loadView() {
    view = UITableView(frame: .zero, style: .insetGrouped)
  }

  override func viewDidLoad() {
    super.viewDidLoad()
    configureNavigationBar()
  }

  override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    configureDataSourceProvider()
  }

  // MARK: - DataSourceProviderDelegate

  func didSelectRowAt(_ indexPath: IndexPath, on tableView: UITableView) {
    let item = dataSourceProvider.item(at: indexPath)

    guard let actionName = item.detailTitle,
          let action = SettingsAction(rawValue: actionName) else {
      // The row tapped has no affiliated action.
      return
    }
    let auth = AppManager.shared.auth()

    switch action {
    case .toggleSecureTokenAPI:
      toggleSecureTokenAPI()
    case .toggleIdentityTokenAPI:
      toggleIdentityTokenAPI()
    case .toggleActiveApp:
      AppManager.shared.toggle()
    case .toggleAccessGroup:
      toggleAccessGroup()
    case .setAuthLanguage:
      setAuthLanguage()
    case .useAppLanguage:
      auth.useAppLanguage()
    case .toggleAPNSToken:
      clearAPNSToken()
    case .toggleAppCredential:
      clearAppCredential()
    case .togglePhoneAppVerification:
      guard let settings = auth.settings else {
        fatalError("Unset auth.settings")
      }
      settings.isAppVerificationDisabledForTesting = !settings.isAppVerificationDisabledForTesting
    }
    updateUI()
  }

  // MARK: - Firebase 🔥

  private func toggleIdentityTokenAPI() {
    if IdentityToolkitRequest.host == "www.googleapis.com" {
      IdentityToolkitRequest.setHost("staging-www.sandbox.googleapis.com")
    } else {
      IdentityToolkitRequest.setHost("www.googleapis.com")
    }
  }

  private func toggleSecureTokenAPI() {
    if SecureTokenRequest.host == "securetoken.googleapis.com" {
      SecureTokenRequest.setHost("staging-securetoken.sandbox.googleapis.com")
    } else {
      SecureTokenRequest.setHost("securetoken.googleapis.com")
    }
  }

  private func toggleAccessGroup() {
    do {
      if AppManager.shared.auth().userAccessGroup == nil {
        guard let bundleDictionary = Bundle.main.infoDictionary,
              let group = bundleDictionary["AppIdentifierPrefix"] as? String else {
          fatalError("Configure AppIdentifierPrefix in the plist")
        }
        try AppManager.shared.auth().useUserAccessGroup(group +
          "com.google.firebase.auth.keychainGroup1")
      } else {
        try AppManager.shared.auth().useUserAccessGroup(nil)
      }
    } catch {
      fatalError("Failed to set userAccessGroup with error \(error)")
    }
  }

  func clearAPNSToken() {
    guard let token = AppManager.shared.auth().tokenManager.token else {
      return
    }

    let tokenType = token.type == .prod ? "Production" : "Sandbox"
    let message = "token: \(token.string)\ntype: \(tokenType)"

    let prompt = UIAlertController(title: "Clear APNs Token?", message: message,
                                   preferredStyle: .alert)
    let okAction = UIAlertAction(title: "OK", style: .default) { action in
      AppManager.shared.auth().tokenManager.token = nil
      self.updateUI()
    }
    prompt.addAction(okAction)
    present(prompt, animated: true)
  }

  func clearAppCredential() {
    if let credential = AppManager.shared.auth().appCredentialManager.credential {
      let message = "receipt:\(credential.receipt) secret:\(credential.secret ?? "nil")"
      let prompt = UIAlertController(title: "Clear App Credential", message: message,
                                     preferredStyle: .alert)
      let okAction = UIAlertAction(title: "OK", style: .default) { action in
        AppManager.shared.auth().appCredentialManager.clearCredential()
        self.updateUI()
      }
      prompt.addAction(okAction)
      present(prompt, animated: true)
    }
  }

  private func setAuthLanguage() {
    let prompt = UIAlertController(title: nil, message: "Enter Language Code For Auth:",
                                   preferredStyle: .alert)
    prompt.addTextField()
    let okAction = UIAlertAction(title: "OK", style: .default) { action in
      AppManager.shared.auth().languageCode = prompt.textFields?[0].text ?? ""
      self.updateUI()
    }
    prompt.addAction(okAction)

    present(prompt, animated: true)
  }

  // MARK: - Private Helpers

  private func configureNavigationBar() {
    navigationItem.title = "Settings"
    guard let navigationBar = navigationController?.navigationBar else { return }
    navigationBar.prefersLargeTitles = true
    navigationBar.titleTextAttributes = [.foregroundColor: UIColor.systemOrange]
    navigationBar.largeTitleTextAttributes = [.foregroundColor: UIColor.systemOrange]
  }

  private func configureDataSourceProvider() {
    dataSourceProvider = DataSourceProvider(
      dataSource: settings?.sections,
      emptyStateView: SignedOutView(),
      tableView: tableView
    )
    dataSourceProvider.delegate = self
  }

  private func updateUI() {
    configureDataSourceProvider()
    animateUpdates(for: tableView)
  }

  private func animateUpdates(for tableView: UITableView) {
    UIView.transition(with: tableView, duration: 0.2,
                      options: .transitionCrossDissolve,
                      animations: { tableView.reloadData() })
  }
}

// MARK: - Extending a `AuthSettings` to conform to `DataSourceProvidable`

extension AuthSettings: DataSourceProvidable {
  private var versionSection: Section {
    let items = [Item(title: FirebaseVersion(), detailTitle: "FirebaseAuth")]
    return Section(headerDescription: "Versions", items: items)
  }

  private var apiHostSection: Section {
    let items = [Item(title: IdentityToolkitRequest.host, detailTitle: "Identity Toolkit"),
                 Item(title: SecureTokenRequest.host, detailTitle: "Secure Token")]
    return Section(headerDescription: "API Hosts", items: items)
  }

  private var appsSection: Section {
    let items = [Item(title: AppManager.shared.app.options.projectID, detailTitle: "Active App")]
    return Section(headerDescription: "Firebase Apps", items: items)
  }

  private var keychainSection: Section {
    let items = [Item(title: AppManager.shared.auth().userAccessGroup ?? "[none]",
                      detailTitle: "Current Access Group")]
    return Section(headerDescription: "Keychain Access Groups", items: items)
  }

  func truncatedString(string: String, length: Int) -> String {
    guard string.count > length else { return string }

    let half = (length - 3) / 2
    let startIndex = string.startIndex
    let midIndex = string.index(startIndex, offsetBy: half) // Ensure correct mid index
    let endIndex = string.index(startIndex, offsetBy: string.count - half)

    return "\(string[startIndex ..< midIndex])...\(string[endIndex...])"
  }

  // TODO: Add ability to click and clear both of these fields.
  private var phoneAuthSection: Section {
    let items = [Item(title: APNSTokenString(), detailTitle: "APNs Token"),
                 Item(title: appCredentialString(), detailTitle: "App Credential")]
    return Section(headerDescription: "Phone Auth", items: items)
  }

  func APNSTokenString() -> String {
    guard let token = AppManager.shared.auth().tokenManager.token else {
      return "No APNs token"
    }

    let truncatedToken = truncatedString(string: token.string, length: 19)
    let tokenType = token.type == .prod ? "Production" : "Sandbox"
    return "\(truncatedToken)(\(tokenType))"
  }

  func appCredentialString() -> String {
    if let credential = AppManager.shared.auth().appCredentialManager.credential {
      let truncatedReceipt = truncatedString(string: credential.receipt, length: 13)
      let truncatedSecret = truncatedString(string: credential.secret ?? "", length: 13)
      return "\(truncatedReceipt)/\(truncatedSecret)"
    } else {
      return "No App Credential"
    }
  }

  private var languageSection: Section {
    let languageCode = AppManager.shared.auth().languageCode
    let items = [Item(title: languageCode ?? "[none]", detailTitle: "Auth Language"),
                 Item(title: "Click to Use App Language", detailTitle: "Use App Language")]
    return Section(headerDescription: "Language", items: items)
  }

  private var disableSection: Section {
    guard let settings = AppManager.shared.auth().settings else {
      fatalError("Missing auth settings")
    }
    let disabling = settings.isAppVerificationDisabledForTesting ? "YES" : "NO"
    let items = [Item(title: disabling, detailTitle: "Disable App Verification (Phone)")]
    return Section(headerDescription: "Auth Settings", items: items)
  }

  var sections: [Section] {
    [
      versionSection,
      apiHostSection,
      appsSection,
      keychainSection,
      phoneAuthSection,
      languageSection,
      disableSection,
    ]
  }
}
